docs/glossary.md

```diff
@@ -0,0 +1,101 @@
+2. Unusually tight integration between the evaluator and the environment.
+    Most languages have cleanly separated editor->files->parser->AST->evaluator; in Boxer you'll see direct calls editor<->evaluator in both directions.
+3. Sleight of hand!  A naive evaluator was too slow (at least decades ago), so behind the scenes a lot is optimized & cached.  When not stepping, most of the execution _does_ happen in hidden state, but care was taken to make this equivalent.
+
+## explicit control evaluator
+
+The `boxer-eval` state machine maintains its state in explicit global variables (one can think of them as registers) and stacks.
+Procedure calls in boxer grow boxer's stack, not the underlying Lisp's stack.
+
+This exposes the state for stepper / debugging / error handling.  It also allows some forms of context switching between "processes".
+
+### pdl (from archaic "Push Down List")
+
+This is the main control stack.
+
+It consists of typed frames, each saving/restoring a different subset of state.
+
+### vpdl (from "Value Push Down List")
+
+A stack of values which will become arguments to functions.
+
+The naive "copy and execute" model actually mutates the calling box, replacing calls with their return values, variables with their values etc.
+As an optimization when not stepping, this can mostly be skipped — and VPDL stack holds the intermediate values.
+
+It's also the interface to primitives.  `defboxer-primitive` looks similar to `defun` but constructs a lisp functions that pops arguments from VPDL, and pushes return value to VPDL.
+
+### recursive evaluation
+
+The evaluator calls primitives.  Primitives like `repeat` or `tell` that need to run boxer code can't just call the evaluator (that would grow the lisp stack and would be opaque to the stepper).
+
+`defrecursive-eval-primitive` / `defrecursive-funcall-primitive` macros split each such primitive into a before, after, and cleanup parts, plus a custom PDL frame type.  Between the parts, control returns to the main eval loop, with state (e.g. `repeat`'s loop counter) retained in the PDL frame.
+
+### process
+
+### sfun
+
+### ufun
+
+## vc - virtual copy
+
+## Variables in Boxer
+
+All this applies equally to variable and procedure lookup by name, the rules are the same.
+(Lookup finds a named box; only then you'll see if it's a data or doit box, and what to do with it.
+In this sense, Boxer falls into ["Lisp-1" camp](http://www.nhplace.com/kent/Papers/Technical-Issues.html).)
+
+### Static
+
+Under the "copy and execute" model, a called box is copied into the calling box.
+Names are looked up directly in the hierarchy of boxes containing the current box.
+This naturally resulting semantic is dynamic binding — free variables are can be bound by the caller.
+
+That's what "static" lookup refers to — search up the hierarchy of materialized boxes.
+
+(The term has nothing to do with C/Java meaning of "static" lifetime.)
+
+### Dynamic
```

Comment:
yes, it's totally confusing :100:.  These are terms used inside the code, e.g. src/evaluator/bind.lisp so I can't change them, but let me see what I can do about explicitly pointing out the overloaded "dynamic" word...

It seems the code mostly talks about _static/dynamic variables_ which are an internal optimization distinction (but both implement the _dynamic scope_ semantics).  
I'm trying to stress this being _static/dynamic lookup_ here which makes more sense than actually ascribing it to the variables.

(lexical scoping is possible via ports.  I guess I better add a mention to that too.)

"Structural scoping" is a nice term.  But not sure it helps given existing terminology in code, and "dynamic scoping" already being a globally established term.  
I'll try to sneak in "structural" not as a term but as an explanation.

Reply:
There are actually 3 things I better separate:

1. Boxer's surrogate model, which prescribes structural lookup in containing boxes.
  This is what a user is supposed to learn, and can observe when stepping.
2. The variable semantics that naturally fall out from this execution model are dynamic scoping.  This is what a computer scientist would notice.
3. The implementation cheats on implementing the model!  
   - Dynamic variables stack(?) holds bindings from copy boxes that were not created.
   - "Static lookup continues where dynamic lookup leaves off" (Klotz1989) and does follow the structural model.

In a sense, box copying was optimized twice:  
Virtual copy mechanism makes copies cheaper but the code still accesses them as totally real boxes.
But then boxes may not be copied at all, and their would-be state is spread around (dynamic variables stack instead of named sub-boxes + temp data on VPDL stack instead of mutating the + return pointer on PDL stack instead of parent box relation, probably more I'm missing...).  In a way you could conceptualize these as representing an even cheaper "phantom box" — but this time the code does NOT pretend it's a box at all, it refers to a bunch of separate ad-hoc mechanisms. Oh well.

relevant doc: Klotz1989 pp. 32–35 sections 5.2.2 Stacks -> Dynamic Variables Stack + 5.2.3 Variable Reference

Reply:
TIL the word "dynamic binding" is somewhat overloaded and may be understood as [late binding](https://en.wikipedia.org/wiki/Late_binding) at runtime, as opposed to early/static binding at compile time.  
Well, boxer totally does late binding, and the concepts do correlate (it's hard to implement dynamic scoping with early binding) but that's yet another layer of confusion I don't need here.  
=> Gonna stick to "dynamic scoping".

Reply:
@sgithens @amb26 rewrote this section significantly, please review.